lib/ts/recipe/session/recipeImplementation.ts

```diff
@@ -162,38 +171,59 @@ export default function getRecipeInterface(
+                options !== undefined && typeof options !== "boolean" && options.sessionRequired === false;
+            const gotRID = frontendHasInterceptor(req);
+            const method = req.getMethod();
+
+            logDebugMessage("getSession: optional validation: " + sessionOptional);
+            logDebugMessage("getSession: rid in header: " + gotRID);
+            logDebugMessage("getSession: request method: " + method);
             let doAntiCsrfCheck = options !== undefined ? options.antiCsrfCheck : undefined;
-            const transferMethod = config.getTokenTransferMethod({ req, userContext });
+            const preferredTransferMethod = config.getTokenTransferMethod({ req, userContext });
+
+            let transferMethod = preferredTransferMethod === "MISSING_AUTH_HEADER" ? "header" : preferredTransferMethod;
 
-            let accessToken = getToken(config, req, "access", userContext, transferMethod);
+            let accessToken = getToken(req, "access", transferMethod);
             if (accessToken === undefined) {
+                const fallbackMethod = transferMethod === "cookie" ? "header" : "cookie";
                 // We are checking if we could've gone ahead with validation if the transferMethod was different
                 // However, we don't want to do the fallback here, but force a call to refresh
                 // This is done to ensure that the browsers update the transfermethod in a timely manner (basically on the next API call)
                 // instead of waiting for the session to expire
-                accessToken = getToken(
-                    config,
-                    req,
-                    "access",
-                    userContext,
-                    transferMethod === "cookie" ? "header" : "cookie"
-                );
+                accessToken = getToken(req, "access", fallbackMethod);
                 if (accessToken !== undefined) {
-                    logDebugMessage(
-                        "getSession: Returning try refresh token because the access token was not sent as a " +
-                            transferMethod
-                    );
+                    if (preferredTransferMethod === "MISSING_AUTH_HEADER") {
+                        transferMethod = fallbackMethod;
+                        logDebugMessage(
+                            `getSession: falling back to use ${fallbackMethod} as an authentication-method`
+                        );
+                    } else {
+                        logDebugMessage(
+                            `getSession: returning TRY_REFRESH_TOKEN because preferred auth-mode doesn't allow fallback and access token was sent with the wrong method`
+                        );
+                        throw new STError({
+                            message: "access token sent with the wrong method. Please call the refresh API",
+                            type: STError.TRY_REFRESH_TOKEN,
+                        });
```

Comment:
Shouldn't we actually send this is the input access token is valid? What if that access token has expired, or it's an access token for another system etc.. we want to try refresh only if it's our access token otherwise we want to throw unauthorised (or return undefined in case of optional session verification).

This is useful in case of an API can be called by two types of clients - one that uses supertokens session and one that doesn't. In this case, the API would need to do getSession with optional true and if the auth mode is cookie based, but the input is an authorisation token, containing the other system's access token, then we want to let it pass.